src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -0,0 +1,382 @@
+        {
+            lock (_sync)
+            {
+                if (_currentFlushTcs == null && _head == null)
+                {
+                    return _innerPipeWriter.GetMemory(sizeHint);
+                }
+
+                AllocateMemoryUnsynchronized(sizeHint);
+                return _tailMemory;
+            }
+        }
+
+        public override Span<byte> GetSpan(int sizeHint = 0)
+        {
+            lock (_sync)
+            {
+                if (_currentFlushTcs == null && _head == null)
+                {
+                    return _innerPipeWriter.GetSpan(sizeHint);
+                }
+
+                AllocateMemoryUnsynchronized(sizeHint);
+                return _tailMemory.Span;
+            }
+        }
+
+        public override void Advance(int bytes)
+        {
+            lock (_sync)
+            {
+                if (_currentFlushTcs == null && _head == null)
+                {
+                    _innerPipeWriter.Advance(bytes);
+                    return;
+                }
+
+                if ((uint)bytes > (uint)_tailMemory.Length)
+                {
+                    ThrowArgumentOutOfRangeException(nameof(bytes));
+                }
+
+                _tailBytesBuffered += bytes;
+                _bytesBuffered += bytes;
+                _tailMemory = _tailMemory.Slice(bytes);
+                _bufferedWritePending = false;
+            }
+        }
+
+        // This is not exposed to end users. Throw so we find out if we ever start calling this.
+        public override void CancelPendingFlush()
+        {
+            // If we wanted, we could propagate IsCanceled when we do multiple flushes in a loop.
+            // If FlushResult.IsCanceled is true with more data pending to flush, we could complete _currentFlushTcs with canceled flush task,
+            // but rekick the FlushAsync loop.
+            throw new NotImplementedException();
+        }
+
+        // This is not exposed to end users. Throw so we find out if we ever start calling this.
+        public override void Complete(Exception exception = null)
+        {
+            // If we wanted, we could store the complete exception or some sentinel exception instance in a field
+            // if a flush was ongoing. We'd just call the inner Complete() method after the flush loop ended.
+
+            // To simply ensure everything gets returned after the PipeWriter is left in some unknown state (say GetMemory() was
+            // called but not Advance(), or there's a flush pending), just call Abort().
+            throw new NotImplementedException();
+        }
+
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            lock (_sync)
+            {
+                if (_currentFlushTcs != null)
+                {
+                    return new ValueTask<FlushResult>(_currentFlushTcs.Task);
```

Comment:
> Maybe we should have this throw.

This will likely throw in the future for any PipeWriter that we decide to expose. This PipeWriter is strictly internal and used by Http2FrameWriter which expects to be able to append more data while another stream is in the middle of flushing the connection-level pipe.

Reply:
Yeah this comment was made before I realized the design